### ISSUE TITLE
Fonts look different in some flatpaked Java programs

### ISSUE THREAD

**Issue by luni3359** [flatpak/flatpak#4122]
## Linux distribution and version
KDE Neon User Edition 5.20 x86_64
## Flatpak version
1.10.1
## Description of the problem
Fonts are showing aliased in some flatpaked Java programs. Running these jars without a sandboxed environment results in proper, smoothed-out fonts.

First reported in the repo where I saw this problem occur, but down the line the maintainer told me the issue wasn't on their flatpak, not on their machine.
guihkx/com.teamshiginima.ShiginimaLauncher#1

Here's a picture of how the fonts differ (On `org.jdownloader.JDownloader`):
![Sidebyside](https://user-images.githubusercontent.com/16832398/107577558-33137600-6bb8-11eb-8342-c4d672133fbb.png)

Another picture. This time the proportions of the windows look different. Top is native, bottom flatpak.
![image](https://user-images.githubusercontent.com/16832398/107579196-60612380-6bba-11eb-8980-6f2b492bbc1d.png)

The maintainer told me they didn't experience issues on their system running GNOME.

## Steps to reproduce
1. Run program
2. Program fonts are different

**Comment by zocker-160** [comment]
any updates on this?

I am experiencing exactly the same issue on Kubuntu with every Flatpaked Java application.

**Comment by luni3359** [comment]
It's still bad. I think there's a fonts issue with flatpak as a whole because if you try to run any flatpak with a Japanese locale, a different font is used which makes it impossible to read sometimes because characters are replaced with a square. 
Here's some proof.
![Broken font](https://user-images.githubusercontent.com/16832398/189465288-d004aecc-2722-483a-8f46-e3e0353f75b3.png)

**Comment by zocker-160** [comment]
yeah this is a dealbreaker, I think this should be reported to the openjdk extension repo if not already done

**Comment by soredake** [comment]
Install xdg-desktop-portal-gtk and do `sudo flatpak override --filesystem=xdg-config/fontconfig:ro `

**Comment by zocker-160** [comment]
thanks you for this suggestion, sadly it does not solve the issue.

The text does look different, so it does _something_, but it is nowhere near where it should be.
here an example:

native:
![Screenshot_20220910_200236](https://user-images.githubusercontent.com/36563538/189496226-82f01b94-2173-4b61-973b-5ea50e378839.png)

flatpak default permissions:
![Screenshot_20220910_200433](https://user-images.githubusercontent.com/36563538/189496253-bcfbe356-8279-466f-a6f1-8edaaa264576.png)

flatpak with `--filesystem=xdg-config/fontconfig:ro`:
![Screenshot_20220910_200142](https://user-images.githubusercontent.com/36563538/189496271-612ae9a9-4be2-4aa9-97c8-bfa80f339720.png)

**Comment by soredake** [comment]
@zocker-160 can you check if that gtk portal is running?

**Comment by zocker-160** [comment]
@soredake I have `xdg-desktop-portal-gtk` and QDbusviewer does show me an `org.freedesktop.impl.portal.desktop.gtk` entry, so I think it is running and working

**Comment by soredake** [comment]
@zocker-160 to clarify, your problem is the used font in jdownloader flatpak is different from the host?

**Comment by zocker-160** [comment]
@soredake not only JDownloader, it is true for any Java application running inside Flatpak.

As you can see on the images above, it is not only a different font, but also the anti aliasing does not seem to work properly. In terms of the last image, even the text color is different.

**Comment by soredake** [comment]
@zocker-160 there's three bugs there
1) KDE don't write AA settings to fontconfig file https://bugs.kde.org/show_bug.cgi?id=416140, you need to disable AA in kde settings and enabled it, then press apply (this need to be done once after installing system until it's fixed)
2) Flatpak does not read your host font settings, it uses font settings from freedesktop-sdk by default https://github.com/flatpak/flatpak/issues/1563, you can fix this by doing (example for noto fonts and hack (as they are used in kubuntu by default), if you want another fonts to be used, you need copy .conf files from /etc/fonts/conf.d):
```
mkdir $HOME/.config/fontconfig/conf.d
cp /etc/fonts/conf.d/*noto* /etc/fonts/conf.d/*hack* "$HOME/.config/fontconfig/conf.d"
```
3) Flatpak does not read your user fontconfig settings https://github.com/flatpak/flatpak/issues/3947, to fix this `flatpak override --user --filesystem=xdg-config/fontconfig:ro`, this will give access to `~/.config/fontconfig/fonts.conf` and `~/.config/fontconfig/conf.d` (you can copy config for fonts you want to be used in apps)

After doing all those three, fonts should look as they needs to be.

Additionally you can try `flatpak override --user --env=_JAVA_OPTIONS='-Dawt.useSystemAAFontSettings=on -Dswing.aatext=true'` or `flatpak override --user --env=_JAVA_OPTIONS='-Dawt.useSystemAAFontSettings=on'`

**Comment by zocker-160** [comment]
@soredake thank you a lot for looking into this, I highly appreciate it.

All 3 issues you listed to me sound like missing file permission, but I tried with `--filesystem=host` which should include everything above but it still looks the same.

I also copied the entire `conf.d` folder from `/etc/fonts` to `~/.config/fontconfig/conf.d` with still no difference.

`-Dswing.aatext=true` does actually do something, it looks aliased now which is something I guess.

To me this sounds like some fundamental issue with Flatpak, since with full access to the host files, the application should have everything it needs.

**Comment by soredake** [comment]
@zocker-160 `--filesystem=host` will not give access to /etc as flatpak have it's own /etc.

Can you show me output of `systemctl --user status xdg-desktop-portal-gtk.service` and `cat ~/.config/fontconfig/fonts.conf`

**Comment by zocker-160** [comment]
@soredake 

> `--filesystem=host` will not give access to /etc as flatpak have it's own /etc.

understood, but I copied the entire folder to `~/.config/fontconfig`

**Comment by zocker-160** [comment]
oops I pressed enter too soon, will post outputs any moment

**Comment by soredake** [comment]
> @soredake
> 
> > `--filesystem=host` will not give access to /etc as flatpak have it's own /etc.
> 
> understood, but I copied the entire folder to `~/.config/fontconfig`

I know, just saying that sharing host `/etc` is not supported https://github.com/flatpak/flatpak/issues/4525

**Comment by zocker-160** [comment]
here my outputs:
```
$ systemctl --user status xdg-desktop-portal-gtk.service
● xdg-desktop-portal-gtk.service - Portal service (GTK/GNOME implementation)
     Loaded: loaded (/usr/lib/systemd/user/xdg-desktop-portal-gtk.service; static)
     Active: active (running) since Sun 2022-09-11 10:25:52 CEST; 6h ago
   Main PID: 3025 (xdg-desktop-por)
      Tasks: 4 (limit: 38257)
     Memory: 4.9M
        CPU: 117ms
     CGroup: /user.slice/user-1000.slice/user@1000.service/app.slice/xdg-desktop-portal-gtk.service
             └─3025 /usr/libexec/xdg-desktop-portal-gtk

Sep 11 10:25:52 Zocker-KDE systemd[2838]: Starting Portal service (GTK/GNOME implementation)...
Sep 11 10:25:52 Zocker-KDE systemd[2838]: Started Portal service (GTK/GNOME implementation).
```
```
$ cat ~/.config/fontconfig/fonts.conf
<?xml version='1.0'?>
<!DOCTYPE fontconfig SYSTEM 'fonts.dtd'>
<fontconfig>
 <dir>~/.fonts</dir>
 <match target="font">
  <edit mode="assign" name="hinting">
   <bool>true</bool>
  </edit>
 </match>
 <match target="font">
  <edit mode="assign" name="hintstyle">
   <const>hintslight</const>
  </edit>
 </match>
 <match target="font">
  <edit mode="assign" name="rgba">
   <const>rgb</const>
  </edit>
 </match>
 <match target="font">
  <edit mode="assign" name="antialias">
   <bool>true</bool>
  </edit>
 </match>
</fontconfig>
```

> I know, just saying that sharing host /etc is not supported https://github.com/flatpak/flatpak/issues/4525

understood yeah you are right, good to know and good point

**Comment by soredake** [comment]
@zocker-160 yeah, everything is setup correctly. If your fonts are different in flatpak'ed java apps that can mean three things:
1) Maybe internal flatpak fontconfig files overriding your fonts (as file which loads your settings are lower priority than flatpak internal configs)
![изображение](https://user-images.githubusercontent.com/5204968/189536441-cf0ab4e8-1446-46c7-8cd6-b5e8275496e9.png)

2) Probably different java versions render/read font config differently. You can report your problem here https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk17

Jdownloader on host uses 1.8.0 (really old) java, flatpak version uses 17 (recent) java, other flatpaked apps probably use newer java than on host too.

**Comment by zocker-160** [comment]
@soredake I did some further testing with [one of my own applications](https://github.com/zocker-160/AdK-BIN-Editor) which uses Java 17.

And while it does indeed look very similar (way more similar than with a Java 8 or 11 project) to native, it is absolutely not the same.

native:
![Screenshot_20220911_183205](https://user-images.githubusercontent.com/36563538/189538741-4843a9eb-e838-48d6-aede-5710927eb0db.png)

flatpak:
![Screenshot_20220911_183231](https://user-images.githubusercontent.com/36563538/189538759-ef78d51f-6704-4f87-b3b4-ab1e1fb0869b.png)

Now I hear you saying, that one would need to pixel peep in order to see the difference, I agree, BUT this is more a fundamental question, because IMO packaging a software in a different format should absolutely never change how the application looks.

This does not happen for DEB, RPM or Appimage for that matter, so it shouldn't for Flatpak either.

If you want to try yourself, this is my Flatpak manifest for testing:
[com.github.zocker_160.AdK-BIN-Editor.txt](https://github.com/flatpak/flatpak/files/9543067/com.github.zocker_160.AdK-BIN-Editor.txt)

**Comment by soredake** [comment]
@zocker-160 don't forget that flatpak is not only a package format but also have sandboxing (that for now causes problems with integration with system fonts), you can as i said report problem [here](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk17), maybe java sdk maintainer can help you, or complain here https://github.com/flatpak/flatpak/issues/1563
https://github.com/flatpak/flatpak/issues/3947

**Comment by zocker-160** [comment]
> don't forget that flatpak is not only a package format

@soredake yes I know and I highly appreciate the sandboxing as well. I hope the Java SDK maintainers know better what is going on and how to potentially fix it.

Thank you very much for your time and effort to help out with this issue.

**Comment by soredake** [comment]
> 

> native: ![Screenshot_20220911_183205](https://user-images.githubusercontent.com/36563538/189538741-4843a9eb-e838-48d6-aede-5710927eb0db.png)
> 
> flatpak: ![Screenshot_20220911_183231](https://user-images.githubusercontent.com/36563538/189538759-ef78d51f-6704-4f87-b3b4-ab1e1fb0869b.png)

Judging by screenshot it looks like another font is used, so my theory about flatpak internal fonts taking precedence is right, one idea is to remove files from /var/lib/flatpak/<path to runtime's conf.d folder> and leave only 50-flatpak.conf, this will in theory will load only your fonts.

**Comment by zocker-160** [comment]
oh I see, inside the runtime that I am using `org.freedesktop.Platform` there are a lot of fonts placed in `/etc/fonts/conf.d` which are completely different from the ones I have on my host.

How can I get rid of them?

**Comment by soredake** [comment]
@zocker-160 I'm not sure if this will work, but you can try:
1) find `fonts/conf.d` folder of your runtime `find /var/lib/flatpak -type d  -path '*/fonts/conf.d'`
2) delete all files except `50-flatpak.conf` and `50-user.conf `
3) copy all host's `/etc/fonts/conf.d` content to `$HOME/.config/fontconfig/conf.d`

After you've done experimenting you can repair runtimes with `flatpak repair`

**Comment by soredake** [comment]
Or you can change priority of 50 files to 100 to make them load your fonts last overriding flatpak defaults.

**Comment by soredake** [comment]
Another option is to create custom runtime and run your app with `flatpak run --runtime=<your runtime>`
https://github.com/flatpak/flatpak/issues/4042#issue-775530963

**Comment by zocker-160** [comment]
alright so I did some further testing today and [your suggestion](https://github.com/flatpak/flatpak/issues/4122#issuecomment-1243019873) sadly makes things even worse, now the fonts are not only ugly but also very small.

I also tried copying the `/fonts/conf.d` folder from the host into the runtime and that seems to be the most promising so far, it looks the closest to native in terms of font rendering, but the font used is still different.

Maybe I need to copy the actual fonts from the host also into the runtime?

**Comment by zocker-160** [comment]
@soredake I managed success!!!

I am sure this "solution" is ugly as fk and IMO this should be addressed by flatpak devs, no user should have to do any of this.

Here is what I did:

- reset all flatpak runtimes, no changes are needed there (`flatpak repair`)
- `mv ~/.config/fontconfig/fonts.conf ~/.config/fontconfig/fonts.conf.bak`
- `cp -r /etc/fonts ~/.config/fontconfig`
- `cp -r /usr/share/fonts ~/.config/fontconfig`
- add `<dir>~/.config/fontconfig/fonts</dir>` to `fonts.conf`

I think the main issue is, that the Flatpak cannot access the fonts inside `/usr/share/fonts` and `/usr/local/share/fonts`

**Comment by zocker-160** [comment]
adding on overwrite for `/usr/share/fonts` does not work either, the folders have completely different contents

on my host:
```
$ ls -l /usr/share/fonts
insgesamt 36
drwxr-xr-x  2 root root 4096 Aug 13 13:22 cmap
drwxr-xr-x  2 root root 4096 Aug 13 13:22 cMap
drwxr-xr-x  3 root root 4096 Jun 12  2021 eot
drwxr-xr-x  6 root root 4096 Jun 12  2021 opentype
drwxr-xr-x  3 root root 4096 Jun 12  2021 svg
drwxr-xr-x 62 root root 4096 Aug 13 13:34 truetype
drwxr-xr-x  5 root root 4096 Nov  4  2020 type1
drwxr-xr-x  3 root root 4096 Jun 12  2021 woff
drwxr-xr-x  6 root root 4096 Jul 31  2020 X11
```
inside flatpak:
```
$ ls -l /usr/share/fonts/
insgesamt 28
drwxr-xr-x 2 nfsnobody nfsnobody 4096  1. Jan 1970  dejavu
drwxr-xr-x 2 nfsnobody nfsnobody 4096  1. Jan 1970  gnu-free
drwxr-xr-x 2 nfsnobody nfsnobody 4096  1. Jan 1970  google-crosextra-caladea
drwxr-xr-x 2 nfsnobody nfsnobody 4096  1. Jan 1970  google-crosextra-carlito
drwxr-xr-x 2 nfsnobody nfsnobody 4096  1. Jan 1970  liberation-fonts
drwxr-xr-x 2 nfsnobody nfsnobody 4096  1. Jan 1970  noto-emoji
drwxr-xr-x 2 nfsnobody nfsnobody 4096  1. Jan 1970  tex-gyre

```

**Comment by soredake** [comment]
> I think the main issue is, that the Flatpak cannot access the fonts inside `/usr/share/fonts` and `/usr/local/share/fonts`

Your host fonts are mounted `/run/host/fonts/` which should be loaded with right configuration (`/etc/fonts/conf.d/50-
50-flatpak.conf`), but, i'm not an expert with fontconfig.

**Comment by RalfJung** [comment]
FWIW this does not just affect Java applications. I am seeing this in vscodium, where the text in flatpak is somehow wider than in a vscode running in firejail.

**Comment by sblantipodi** [comment]
@RalfJung I would like to make this "fix" available via my manifest and not ask users to launch a command to fix fonts in my app.

I added this
"--filesystem=xdg-config/fontconfig:ro"
to my finish-args 
but it doesn't work.

Is there someone who found a fix ?
